src/charm.py

```diff
@@ -0,0 +1,769 @@
+        )
+        self.framework.observe(
+            self.ingress.on.lb_configured, self._configure_dashboard
+        )
+
+        # Certificates Relation
+        self.ca_client = ca_client.CAClient(self, "certificates")
+        self.framework.observe(
+            self.ca_client.on.ca_available, self._request_certificates
+        )
+        self.framework.observe(
+            self.ca_client.on.tls_server_config_ready,
+            self._enable_ssl_from_relation
+        )
+        self.framework.observe(
+            self.on["certificates"].relation_departed,
+            self._certificates_relation_departed,
+        )
+
+        # Charm Custom Events
+        self.framework.observe(self.on.disable_ssl, self._clean_ssl_conf)
+        self.framework.observe(
+            self.on.enable_ssl_from_config, self._enable_ssl_from_config
+        )
+
+        self._stored.set_default(is_started=False)
+
+    def _request_loadbalancer(self, _event) -> None:
+        """Send request to create loadbalancer"""
+        self.ingress.request_loadbalancer(
+            self.LB_SERVICE_NAME,
+            self.TLS_PORT,
+            self.TLS_PORT,
+            self._get_bind_ip(),
+            'http',
+        )
+
+    def _register_dashboards(self) -> None:
+        """Register all dashboards with grafana"""
+        if not self.unit.is_leader():
+            return  # Do nothing on non leader units.
+
+        for dash_file in self.DASH_DIR.glob("*.json"):
+            self.grafana_dashboard.register_dashboard(
+                dash_file.stem,
+                json.loads(dash_file.read_text()))
+            logging.debug(
+                "register_grafana_dashboard: {}".format(dash_file))
+
+    def _update_legacy_radosgw_creds(self, access_key: str,
+                                     secret_key: str) -> None:
+        """Update dashboard db with access & secret key for rados gateways.
+
+        This method uses the legacy format which only supports one gateway.
+        """
+        self._apply_file_setting('set-rgw-api-access-key', access_key)
+        self._apply_file_setting('set-rgw-api-secret-key', secret_key)
+
+    def _update_multi_radosgw_creds(self, creds: str) -> None:
+        """Update dashboard db with access & secret key for rados gateway."""
+        access_keys = {c['daemon_id']: c['access_key'] for c in creds}
+        secret_keys = {c['daemon_id']: c['secret_key'] for c in creds}
+        self._apply_file_setting(
+            'set-rgw-api-access-key',
+            json.dumps(access_keys))
+        self._apply_file_setting(
+            'set-rgw-api-secret-key',
+            json.dumps(secret_keys))
+
+    def _support_multiple_gateways(self) -> bool:
+        """Check if version of dashboard supports multiple rados gateways"""
+        return ch_host.cmp_pkgrevno('ceph-common', '16.0') > 0
+
+    def _manage_radosgw(self) -> None:
+        """Register rados gateways in dashboard db"""
+        if self.unit.is_leader():
```

Comment:
nit: probably better to test on `if not self.unit.is_leader()` and return to remove  1 level of nestedness